File: compilers_python/lab2-2/data_types.py
import abc
import enum
import typing
from dataclasses import dataclass
from typing import Optional

class TypeKind(enum.Enum):
    Global = 'GLOBAL'
    Local = 'LOCAL'

@dataclass
class Type:
    kind: TypeKind
    name: str

class GlobalType(enum.Enum):
    Integer = Type(TypeKind.Global, 'INTEGER')
    Real = Type(TypeKind.Global, 'REAL')
    Boolean = Type(TypeKind.Global, 'BOOLEAN')

@dataclass
class VarType:
    type: Type
    is_pointer: bool

@dataclass
class VarsDef:
    names: list[str]
    type: VarType

@dataclass
class TypeDef:
    name: str
    parent_type: Optional[Type]
    varDefs: list[VarsDef]

class Statement(abc.ABC):
    pass

@dataclass
class Program:
    type_defs: list[TypeDef]
    var_defs: list[VarsDef]
    statements: list[Statement]

class Expr(abc.ABC):
    pass

@dataclass
class VarChain:
    variables: list[str]

@dataclass
class AssignStatement(Statement):
    variable: VarChain
    expr: Expr

@dataclass
class PointerStatement(Statement):
    pointer_name: str
    variable: VarChain

@dataclass
class CreateStatement(Statement):
    pointer_name: str

@dataclass
class IfStatement(Statement):
    condition: Expr
    then_branch: list[Statement]
    else_branch: list[Statement]

@dataclass
class WhileStatement(Statement):
    condition: Expr
    body: list[Statement]

@dataclass
class VariableExpr(Expr):
    var_name: VarChain

@dataclass
class ConstExpr(Expr):
    value: typing.Any
    type: Type

@dataclass
class BinOpExpr(Expr):
    left: Expr
    op: str
    right: Expr

@dataclass
class UnOpExpr(Expr):
    op: str
    expr: Expr

File: compilers_swift/lab2-4/src/syntax-analyzer/data_types.swift
import Foundation

enum TypeKind {
    
    case global, local
}

class SomeType {
    
    let kind: TypeKind
    let name: String
    
    init(kind: TypeKind, name: String) {
        self.kind = kind
        self.name = name
    }
}

class VarType {
    
    let type: SomeType
    let isPointer: Bool
    
    init(type: SomeType, isPointer: Bool) {
        self.type = type
        self.isPointer = isPointer
    }
}

class VarDef {
    
    let name: String
    let type: VarType
    
    init(name: String, type: VarType) {
        self.name = name
        self.type = type
    }
}

class TypeDef {
    
    let name: String
    let parentType: SomeType?
    let varDefs: [VarDef]
    
    init(name: String, parentType: SomeType? = nil, varDefs: [VarDef]) {
        self.name = name
        self.parentType = parentType
        self.varDefs = varDefs
    }
}

class Statement {}

class Program {
    
    let typeDefs: [TypeDef]
    let varDefs: [VarDef]
    let statements: [Statement]
    
    init(typeDefs: [TypeDef], varDefs: [VarDef], statements: [Statement]) {
        self.typeDefs = typeDefs
        self.varDefs = varDefs
        self.statements = statements
    }
}

class Expr {}

class VarChain {
    
    let variables: [String]
    
    init(variables: [String]) {
        self.variables = variables
    }
}

class AssignStatement: Statement {
    
    let variable: VarChain
    let expr: Expr
    let isDereference: Bool
    
    init(variable: VarChain, expr: Expr, isDereference: Bool) {
        self.variable = variable
        self.expr = expr
        self.isDereference = isDereference
    }
}

class CreateStatement: Statement {
    
    let pointerName: VarChain
    
    init(pointerName: VarChain) {
        self.pointerName = pointerName
    }
}

class IfStatement: Statement {
    
    let condition: Expr
    let thenBranch: [Statement]
    let elseBranch: [Statement]
    
    init(condition: Expr, thenBranch: [Statement], elseBranch: [Statement]) {
        self.condition = condition
        self.thenBranch = thenBranch
        self.elseBranch = elseBranch
    }
}

class WhileStatement: Statement {
    
    let condition: Expr
    let body: [Statement]
    
    init(condition: Expr, body: [Statement]) {
        self.condition = condition
        self.body = body
    }
}

class VariableExpr: Expr {
    
    let varName: VarChain
    
    init(varName: VarChain) {
        self.varName = varName
    }
}

class ConstExpr: Expr {
    
    let value: Any
    let type: SomeType
    
    init(value: Any, type: SomeType) {
        self.value = value
        self.type = type
    }
}

class BinOpExpr: Expr {
    
    let left: Expr
    let op: String
    let right: Expr
    
    init(left: Expr, op: String, right: Expr) {
        self.left = left
        self.op = op
        self.right = right
    }
}

class UnOpExpr: Expr {
    
    let op: String
    let expr: Expr
    
    init(op: String, expr: Expr) {
        self.op = op
        self.expr = expr
    }
}
